File: iOS-MobileO11yDemo/iOS-MobileO11yDemo/Features/CarConnectionAction/Models/CarSoftwareVersion.swift
//
//  CarSoftwareVersion.swift
//  iOS-MobileO11yDemo
//
//  Created by Robert Magnusson on 04.06.24.
//

import Foundation

struct CarSoftwareVersion: CustomStringConvertible {
    let major: Int
    let minor: Int
    let patch: Int
    
    var description: String {
        return "(v\(major).\(minor).\(patch))"
    }
}

class CarSoftwareVersionFactory {
    func getRandomVersion() -> CarSoftwareVersion {
        let initalVersions = [
            CarSoftwareVersion(major: 1, minor: 0, patch: 0),
            CarSoftwareVersion(major: 1, minor: 2, patch: 0),
            CarSoftwareVersion(major: 2, minor: 0, patch: 1),
            CarSoftwareVersion(major: 2, minor: 1, patch: 2),
        ]
        return initalVersions.randomElement()!
    }
    
    func getRandomNextVersion(currentVersion: CarSoftwareVersion) -> CarSoftwareVersion {
        return CarSoftwareVersion(
            major: Int.random(in: currentVersion.major...(currentVersion.major+1)),
            minor: Int.random(in: currentVersion.minor...(currentVersion.minor+1)),
            patch: Int.random(in: (currentVersion.patch+1)...(currentVersion.patch+5))
        )
    }
}

File: Mobiles/flutter/lib/core/car/domain_layer/car_software_version.dart
import 'package:flutter_mobile_o11y_demo/core/utils/random_element.dart';

class CarSoftwareVersion {
  const CarSoftwareVersion({
    required this.major,
    required this.minor,
    required this.patch,
  });

  final int major;
  final int minor;
  final int patch;

  @override
  String toString() {
    return '(v$major.$minor.$patch)';
  }
}

class CarSoftwareVersionFactory {
  CarSoftwareVersion getRandomVersion() {
    const initialVersions = [
      CarSoftwareVersion(major: 1, minor: 0, patch: 0),
      CarSoftwareVersion(major: 1, minor: 2, patch: 0),
      CarSoftwareVersion(major: 2, minor: 0, patch: 1),
      CarSoftwareVersion(major: 2, minor: 1, patch: 2),
    ];
    return initialVersions.getRandomElement();
  }

  CarSoftwareVersion getRandomNextVersion({
    required CarSoftwareVersion currentVersion,
  }) {
    return CarSoftwareVersion(
      major: RandomElementIntX.getRandomInRange(
        currentVersion.major,
        currentVersion.major + 1,
      ),
      minor: RandomElementIntX.getRandomInRange(
        currentVersion.minor,
        currentVersion.minor + 1,
      ),
      patch: RandomElementIntX.getRandomInRange(
        currentVersion.patch,
        currentVersion.patch + 5,
      ),
    );
  }
}
